File: typescript/test/conformanceConditional.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import {
  FormValidator,
  parsePageJson,
  type FormPageDefinition,
  type ValidationError,
} from "../src/index.js";

// Shared conditional-validation fixture (spec/conformance), consumed identically
// by the Dart and Java editions: hidden fields are not validated, `requiredWhen`
// makes a field required, and an unknown mode leaves mode conditions false.
const fixture = JSON.parse(
  readFileSync("../spec/conformance/conditional_validation.json", "utf8"),
);

const key = (e: ValidationError | { field: string; message: string }): string =>
  `${e.field}=${e.message}`;

describe("conformance: conditional validation", () => {
  const page = parsePageJson(
    JSON.stringify(fixture.page),
  ) as FormPageDefinition;
  const validator = new FormValidator();

  for (const c of fixture.cases as any[]) {
    it(c.name, () => {
      const actual = validator.validate(page.form, c.record, c.mode).errors;
      expect(new Set(actual.map(key))).toEqual(
        new Set((c.expected as any[]).map(key)),
      );
    });
  }
});

File: flutter/packages/hatake_yaml/test/conformance_conditional_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:hatake_core/hatake_core.dart';
import 'package:hatake_yaml/hatake_yaml.dart';
import 'package:test/test.dart';

/// Runs the shared conditional-validation fixture
/// (spec/conformance/conditional_validation.json) — the same contract the
/// TypeScript and Java editions implement.
///
/// ここで固定するのは「隠れている項目は検証しない」「`requiredWhen` が成立したら
/// 必須になる」「mode が分からない場所では mode の条件が false になる」の3つ。
void main() {
  final fixture = jsonDecode(
    File('../../../spec/conformance/conditional_validation.json')
        .readAsStringSync(),
  ) as Map<String, dynamic>;

  final page = parsePageJson(jsonEncode(fixture['page'])) as FormPageDefinition;
  final validator = FormValidator();

  group('conformance: conditional validation', () {
    for (final raw in fixture['cases'] as List) {
      final c = raw as Map<String, dynamic>;
      test(c['name'], () {
        final record = (c['record'] as Map).cast<String, Object?>();
        final actual =
            validator.validate(page.form, record, mode: c['mode'] as String?)
                .errors;

        String key(String field, String message) => '$field=$message';
        final expected = [
          for (final e in c['expected'] as List)
            key((e as Map)['field'] as String, e['message'] as String),
        ];
        expect(
          actual.map((e) => key(e.field, e.message)).toSet(),
          expected.toSet(),
          reason: 'errors: ${actual.map((e) => key(e.field, e.message))}',
        );
      });
    }
  });
}
